src/topology/builder.rs

```diff
@@ -120,9 +121,11 @@ pub async fn build_pieces(
 
         let (output, control) = Fanout::new();
 
+        let filtered = input_rx
+            .filter(move |event| filter_event_type(event, input_type))
+            .inspect(|_| emit!(EventProcessed));
```

Comment:
Won't this print before the event is actually processed, rather than after?

Reply:
Yes it will. 

First, about half of transforms were emitting this metric before doing anything with the event, and half after it. 

Now, this more of an issue of what do we want to count as processed. I see two versions:
* Only events that are processed through a proper, non error, code path is to be counted.
* Or, all events being processed are counted, even if we encounter some error with them. 

The second definition is what is being counted in this PR. The most practical reason as to why is because we cant count the first definition on the topology level. If we would count on the output of the transform we would miscount all sampling/filtering/dedup based transforms.

Reply:
As a side note, that's why no `emit_logs` is added to `EventProcessed`.

Reply:
But if the issue is more of just the timing of emit, we can do something about it.

Reply:
I don't see a way of emitting the metric after it was processed. The main reason is that there is currently no way of injecting code to `Trannsform::Task` that would run once per event and after event was processed. 

Any idea?

Reply:
How about changing the logic here to emit two events: before and after? It makes way more sense from the observability standpoint imo, if we apply this logic at the topology layer. I.e. if we ever have a `blackhole` transform, it will show a lot of events went in and none went out. For the `sampler` transform, it would be obvious how many events were filtered out.

Reply:
> I don't see a way of emitting the metric after it was processed. The main reason is that there is currently no way of injecting code to `Trannsform::Task` that would run once per event and after event was processed.
> 
> Any idea?

I'll think about it... Also, maybe @Hoverbear can help?

Reply:
Maybe we can stick it onto `output` somehow?

Reply:
> Maybe we can stick it onto output somehow?

Nope, the number of events going into output doesn't correspond to the number of events that went into the transform.

> How about changing the logic here to emit two events: before and after?

That would be pretty useful, but we can expand with that in a separate PR.

---

I think the difference between the orderings is not relevant. Currently that event doesn't print anything, and if we change that, we can adjust the log message accordingly. And regarding the metric, that should only be observable from the `internal_metrics` source and the api, right? So that metric would only be out of order relative to other metrics emitted by that transform, for which I don't think we need to care.

Reply:
@MOZGIII I would like to get your take on this before merging.

Reply:
> > How about changing the logic here to emit two events: before and after?
> 
> That would be pretty useful, but we can expand with that in a separate PR.

Yes, I think the goal here should be to match the current behavior. We can discuss transitioning to multiple events separately.

Reply:
> And regarding the metric, that should only be observable from the `internal_metrics` source and the api, right? So that metric would only be out of order relative to other metrics emitted by that transform, for which I don't think we need to care.

We also expose those metrics via prometheus sink in k8s, so users will be able to see raw metrics, and we should try and make them consistent if possible.

I agree that we can transition to the before/after in a separate PR - for what it's worth, it'd be a cleaner change and easier to review.